### DISCUSSION TITLE
New implementation for `proxySet` and `proxyMap` for performance?

### DISCUSSION THREAD

**Discussion by dai-shi** [pmndrs/valtio#943] [Ideas]
Currently our `proxySet` and `proxyMap` are just an emulation with plain JS object/array.
It's because Proxy doesn't support Set and Map natively, and our goal was to provide APIs compatible to Set and Map. @fkhadra implemented the current version and it has been well.

Time has passed since then and we reached Valtio v2. I think it's reasonable to reconsider the implementation of `proxySet` and `proxyMap` for performance, which has requested several times. It means it backs up with the native `Set` and `Map` and we trap all methods of them.

Would anyone like to dig into it? It would require a decent knowledge of Valtio and the Set/Map APIs.

**Comment by overthemike** [comment]
I'll try dig into it If I can get the time to. Don't want to promise anything yet though. I'm not a performance specialist.

Do we have any benchmark goals or anything or are we just trying to improve the current implementation?

**Reply by dai-shi** [reply]
It would be nice if we could eventually measure the performance, but for the start, my interest is whether we can use the native Set/Map internally.

**Reply by overthemike** [reply]
Can we not just extend Map/Set and overwrite just the custom methods? Is this what you mean?

```js
class proxyMap extends Map {
  constructor() {
    super()
  }

  get(key) {
     console.log('getting key', key)
     // do proxy stuff as needed and return
     return super.get(key)
  }
  
  set(key, value){ 
      console.log('setting key', key, value)
      // do proxy stuff
      return super.set(key, value)
  }
}

const m = new proxyMap()
m.set('m', 'mval') 
m.get('m')            

/******** Output
setting key m mval
getting key m
'mval'
*********/
```

**Reply by dai-shi** [reply]
It's an implementation detail. I'm fine with either way, if it works. There might be a risk to forget to override some methods though.

**Reply by fkhadra** [reply]
Nice, I'll probably give it a shot on my side as well. I remember that back in time they were some limitations as we wanted to fully emulate map/set behavior.

**Reply by dai-shi** [reply]
@fkhadra Hi! That would be nice. Yeah, I remember we had a discussion about some limitations. BTW, do you use Discord?

**Reply by fkhadra** [reply]
Hey, yes I do I'll dm you after work. I started to experiment a bit on my side.

**Comment by qingzhoufeihu** [comment]
This should be what proxy from valtio should do, why should we get two separate ones to deal with it?

**Reply by dai-shi** [reply]
because it's not easy to implement. we also want to reduce the code and keep the simplicity of the core.

**Comment by dai-shi** [comment]
#965 merged. :tada:
